chore: Add SecretManager service regional code samples

YashSaraf11 opened the pull request:
### Changes
* Added code samples for regional APIs.
* Added test files for the regional code samples added.
* Added `RegionalSecretManagerFixture.cs` file for testing the regional code samples added.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/AccessRegionalSecretVersionTests.cs:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2020 Google LLC
```
Update copyright year

YashSaraf11 replied:
Sure. I have updated everything to 2024, as per the file creation year.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/AddRegionalSecretVersionTests.cs:
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2020 Google LLC
```
Update copyright year

YashSaraf11 replied:
Sure. I have updated everything to 2024, as per the file creation year.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/CreateRegionalSecretTests.cs:
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2020 Google LLC
```
Update copyright year in all the files

YashSaraf11 replied:
Sure. I have updated everything to 2024, as per the file creation year.

vipul7499 commented on secretmanager/api/SecretManager.Samples/EnableRegionalSecretVersion.cs:
```diff
@@ -0,0 +1,44 @@
+        }.Build();
+
+        // Build the resource name.
+        SecretVersionName secretVersionName = SecretVersionName.FromProjectLocationSecretSecretVersion(projectId, locationId, secretId, secretVersionId);
```
Avoid lengthy lines, break them down, it increases readability.

YashSaraf11 replied:
I have used the `dotnet format` and the breakup was automatically adjusted as per the formatting rules set. I will try to change the formatting rules if possible or manual break the lines.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/DeleteRegionalSecretWithEtagTests.cs:
```diff
@@ -0,0 +1,50 @@
+        Secret secret = client.GetSecret(secretName);
+        string updatedEtag = secret.Etag;
+
+        _sample.DeleteRegionalSecretWithEtag(projectId: secretName.ProjectId, locationId: secretName.LocationId, secretId: secretName.SecretId, etag: updatedEtag);
```
Break down such lines into smaller ones to increase readability

YashSaraf11 replied:
I have used the `dotnet format` and the breakup was automatically adjusted as per the formatting rules set. I will try to change the formatting rules if possible or manual break the lines.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/DisableRegionalSecretVersionWithEtagTests.cs:
```diff
@@ -0,0 +1,51 @@
+        SecretName secretName = _fixture.SecretWithVersions.SecretName;
+        SecretVersionName secretVersionName = _fixture.SecretVersionToDisable.SecretVersionName;
+
+        // Create the Secret Manager Client with the regional endpoint.
```
Update this comment to 
"Create the Regional Secret Manager Client"
Apply this everywhere.

YashSaraf11 replied:
Sure, have made this change everywhere possible.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/ListRegionalSecretVersionsWithFilterTests.cs:
```diff
@@ -0,0 +1,42 @@
+
+        _sample.ListRegionalSecretVersionsWithFilter(
+            projectId: secretName.ProjectId, locationId: secretName.LocationId, secretId: secretName.SecretId, filter: filter);
+    }
```
Please add some assert statements

YashSaraf11 replied:
Have added assert statements here as per the request.

YashSaraf11 replied:
I will update the test.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/ListRegionalSecretsTests.cs:
```diff
@@ -0,0 +1,38 @@
+    {
+        SecretName secretName = _fixture.Secret.SecretName;
+        _sample.ListRegionalSecrets(projectId: secretName.ProjectId, locationId: secretName.LocationId);
+    }
```
Please add some assert statements

YashSaraf11 replied:
Have added assert statements here as per the request.

YashSaraf11 replied:
I will update the test file.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/ListRegionalSecretsWithFilterTests.cs:
```diff
@@ -0,0 +1,41 @@
+        string filter = "name:my-secret";
+
+        _sample.ListRegionalSecretsWithFilter(projectId: secretName.ProjectId, locationId: secretName.LocationId, filter: filter);
+    }
```
Please add some assert statements in all the tests

YashSaraf11 replied:
Have added assert statements here as per the request.

YashSaraf11 replied:
I will update the test file.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/RegionalSecretManagerFixture.cs:
```diff
@@ -0,0 +1,147 @@
+
+    public RegionalSecretManagerFixture()
+    {
+        Console.WriteLine("Hello World!");
```
Remove unnecessary print statements

YashSaraf11 replied:
Have removed this code line, thanks.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/RegionalSecretManagerFixture.cs:
```diff
@@ -0,0 +1,147 @@
+            throw new Exception("missing GOOGLE_PROJECT_ID");
+        }
+
+        // Get LocationId (e.g., "us-central1")
+        LocationId = Environment.GetEnvironmentVariable("GOOGLE_CLOUD_LOCATION") ?? "us-west1";
```
Please make sure both the locations are same

YashSaraf11 replied:
Sure, that was just an example, will update all to `us-west1`.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/RegionalSecretManagerFixture.cs:
```diff
@@ -0,0 +1,147 @@
+        SecretToDelete = CreateSecret(RandomId());
+        SecretToDeleteWithEtag = CreateSecret(RandomId());
+        SecretWithVersions = CreateSecret(RandomId());
+        SecretForQuickstartName = SecretName.FromProjectLocationSecret(ProjectId, LocationId, RandomId());
```
Will this secret be eventually cleaned up as well?

YashSaraf11 replied:
I had missed deleting one of them, have fixed this now to delete and cleanup all of them.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/RegionalSecretManagerFixture.cs:
```diff
@@ -0,0 +1,147 @@
+        {
+            client.DeleteSecret(name);
+        }
+        catch (Grpc.Core.RpcException e) when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
+        {
+            // Ignore error - secret was already deleted
+        }
```
Add some log here

YashSaraf11 replied:
Have added an error console message to be printed.

vipul7499 commented on secretmanager/api/SecretManager.Samples/UpdateRegionalSecretWithAlias.cs:
```diff
@@ -0,0 +1,48 @@
+        Secret secret = new Secret
+        {
+            SecretName = SecretName.FromProjectLocationSecret(projectId, locationId, secretId),
+            VersionAliases = { ["test"] = 1 }
```
What does this signify, add a comment here, and why is this hardcoded?

YashSaraf11 replied:
Have added comments to signify the detail, and this hardcoded just as an example of adding alias to the secret version.

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/UpdateRegionalSecretWithAliasTests.cs:
```diff
@@ -0,0 +1,39 @@
+    {
+        SecretName secretName = _fixture.Secret.SecretName;
+        Secret result = _sample.UpdateRegionalSecretWithAlias(projectId: secretName.ProjectId, locationId: secretName.LocationId, secretId: secretName.SecretId);
+        Assert.Equal(1, result.VersionAliases["test"]);
```
What does this signify?

YashSaraf11 replied:
This is for the test, that the secret version "1" should have alias of "test".

vipul7499 commented on secretmanager/api/SecretManager.Samples.Tests/UpdateRegionalSecretWithEtagTests.cs:
```diff
@@ -0,0 +1,49 @@
+        }.Build();
+
+        Secret secret = client.GetSecret(secretName);
+        string updatedEtag = secret.Etag;
```
Why the updated etag same as original etag?

YashSaraf11 replied:
Its an mistake on the variable name, I will update it to etag for clearing the misconception. Thanks for pointing it out.

vipul7499 commented on secretmanager/api/SecretManager.Samples/ListRegionalSecretVersions.cs:
```diff
@@ -0,0 +1,45 @@
+        // Call the API.
+        foreach (SecretVersion secretVersion in client.ListSecretVersions(secretName))
+        {
+            // ...
```
Please complete this

YashSaraf11 replied:
Have added print statements here as per the request.

vipul7499 commented on secretmanager/api/SecretManager.Samples/ListRegionalSecretVersionsWithFilter.cs:
```diff
@@ -0,0 +1,52 @@
+        // Call the API.
+        foreach (SecretVersion secretVersion in client.ListSecretVersions(request))
+        {
+            // ...
```
Please complete this

YashSaraf11 replied:
Have added console statements here as per the request.

vipul7499 commented on secretmanager/api/SecretManager.Samples/ListRegionalSecrets.cs:
```diff
@@ -0,0 +1,42 @@
+        // Call the API.
+        foreach (Secret secret in client.ListSecrets(locationName))
+        {
+            // ...
```
Please complete this

YashSaraf11 replied:
Have added print statements here as per the request.

vipul7499 commented on secretmanager/api/SecretManager.Samples/ListRegionalSecretsWithFilter.cs:
```diff
@@ -0,0 +1,50 @@
+        // Call the API.
+        foreach (Secret secret in client.ListSecrets(request))
+        {
+            // ...
```
Please complete this

YashSaraf11 replied:
Have added assert statements here as per the request.

vipul7499 commented on secretmanager/api/SecretManager.Samples/RegionalQuickstart.cs:
```diff
@@ -0,0 +1,68 @@
+        AccessSecretVersionResponse result = client.AccessSecretVersion(createdVersion.SecretVersionName);
+
+        // Print the results
+        //
```
Why this extra line?

YashSaraf11 replied:
Have removed this extra line.

vipul7499 commented on secretmanager/api/SecretManager.Samples/UpdateRegionalSecretWithAlias.cs:
```diff
@@ -0,0 +1,48 @@
+        Secret secret = new Secret
+        {
+            SecretName = SecretName.FromProjectLocationSecret(projectId, locationId, secretId),
+            VersionAliases = { ["test"] = 1 }
```
What does this signifiy?

YashSaraf11 commented on secretmanager/api/SecretManager.Samples.Tests/ListRegionalSecretVersionsTests.cs:
```diff
@@ -0,0 +1,39 @@
+        SecretName secretName = _fixture.Secret.SecretName;
+        _sample.ListRegionalSecretVersions(
+            projectId: secretName.ProjectId, locationId: secretName.LocationId, secretId: secretName.SecretId);
+    }
```
Have added assert statements here as per the request.

amanda-tarafa commented on secretmanager/api/SecretManager.Samples.Tests/DisableRegionalSecretVersionWithEtagTests.cs:
```diff
@@ -0,0 +1,56 @@
+          secretVersionId: secretVersionName.SecretVersionId,
+          etag: updatedEtag
+        );
+        Assert.Equal(SecretVersion.Types.State.Disabled, secretVersion.State);
```
And you are leaving this disabled for other tests to fail.

YashSaraf11 replied:
Have created a different secret in the specific test file and also disposes of the created resource in the same test file to prevent the test suite to get flaky.

amanda-tarafa commented on secretmanager/api/SecretManager.Samples.Tests/DestroyRegionalSecretVersionWithEtagTests.cs:
```diff
@@ -0,0 +1,49 @@
+    [Fact]
+    public void DestroysRegionalSecretVersionsWithEtag()
+    {
+        SecretName secretName = _fixture.SecretWithVersions.SecretName;
```
What do you need this for?

YashSaraf11 replied:
Have created a different secret in the specific test file and also disposes of the created resource in the same test file to prevent the test suite to get flaky.

amanda-tarafa commented on secretmanager/api/SecretManager.Samples.Tests/DeleteRegionalSecretTests.cs:
```diff
@@ -0,0 +1,50 @@
+    public void DeletesRegionalSecrets()
+    {
+        SecretName secretName = _fixture.SecretToDelete.SecretName;
+        sss_sample.DeleteRegionalSecret(
```
Typo, this doesn't build.

YashSaraf11 replied:
Have fixed this typo.

amanda-tarafa reviewed (changes requested):
Please rebase this PR on main as at the moment your branch is outdated and this PR is including unrelated Secret Manager changes.

There are several comments that although left in one or two files only, should be addressed everywhere.

I aslo have some concerns about:

- assumptions that tests run in a given order
- assumptions that tests do not run in parallel
- reuse of fxiture resources in tests that modify said resources

All of these may be addressed by creating specific resources for each of the tests that modify the resources. And ideally those should be set up and deleted on each of the tests using the fixture methods as helpers. Shares resources in the fixture should only be used for tests that do not modify those resources. I know some of this was inspired by the existing global samples, but, although those are not perfect, there are way less of them that are modifying resources, and certainly not relying on the eTag for optimistic concurrent updates. This PR as is will be flaky.

Also, the code doesn't currently build, there's a type in a variable name.

YashSaraf11 commented on secretmanager/api/SecretManager.Samples/UpdateRegionalSecret.cs:
```diff
@@ -0,0 +1,51 @@
+    public Secret UpdateRegionalSecret(
+      string projectId = "my-project",
+      string locationId = "my-location",
+      string secretId = "my-secret"
+    )
```
Thanks for catching that! It looks like the lint and formatting tool automatically made these changes.  I'll investigate further and adjust the code to ensure everything passes both the formatting checks and the functional tests. I'll update the PR accordingly.

YashSaraf11 commented on secretmanager/api/SecretManager.Samples.Tests/RegionalSecretManagerFixture.cs:
```diff
@@ -0,0 +1,148 @@
+    public Secret CreateSecret(string secretId)
+    {
+        // Create the Regional Secret Manager Client.
+        SecretManagerServiceClient client = new SecretManagerServiceClientBuilder
```
Have accomodated this change, thanks for pointint it out.

YashSaraf11 commented:
> Please rebase this PR on main as at the moment your branch is outdated and this PR is including unrelated Secret Manager changes.
> 
> There are several comments that although left in one or two files only, should be addressed everywhere.
> 
> I aslo have some concerns about:
> 
> * assumptions that tests run in a given order
> * assumptions that tests do not run in parallel
> * reuse of fxiture resources in tests that modify said resources
> 
> All of these may be addressed by creating specific resources for each of the tests that modify the resources. And ideally those should be set up and deleted on each of the tests using the fixture methods as helpers. Shares resources in the fixture should only be used for tests that do not modify those resources. I know some of this was inspired by the existing global samples, but, although those are not perfect, there are way less of them that are modifying resources, and certainly not relying on the eTag for optimistic concurrent updates. This PR as is will be flaky.
> 
> Also, the code doesn't currently build, there's a type in a variable name.

Have made the changes in the file as per the suggestion, thanks for pointing it out.

YashSaraf11 commented on secretmanager/api/SecretManager.Samples.Tests/EnableRegionalSecretVersionWithEtagTests.cs:
```diff
@@ -0,0 +1,56 @@
+            Endpoint = $"secretmanager.{secretName.LocationId}.rep.googleapis.com"
+        }.Build();
+
+        SecretVersion disabledSecretVersion = client.DisableSecretVersion(secretVersionName);
```
Have created a different secret in the specific test file and also disposes of the created resource in the same test file to prevent the test suite to get flaky.

YashSaraf11 commented on secretmanager/api/SecretManager.Samples.Tests/ListRegionalSecretsWithFilterTests.cs:
```diff
@@ -0,0 +1,55 @@
+    [Fact]
+    public void ListsRegionalSecretsWithFilter()
+    {
+        // Redirect console output to a StringWriter
+        StringWriter stringWriter = new StringWriter();
+        Console.SetOut(stringWriter);
```
Have fixed this and used different assert statements to verify the working instead of redirecting the output.

YashSaraf11 commented on secretmanager/api/SecretManager.Samples/IamRegionalRevokeAccess.cs:
```diff
@@ -0,0 +1,58 @@
+            ResourceAsResourceName = secretName,
+        });
+
+        // Remove the user to the list of bindings.
```
Have accomodated this change, thanks.

YashSaraf11 commented on secretmanager/api/SecretManager.Samples.Tests/IamRegionalRevokeAccessTests.cs:
```diff
@@ -0,0 +1,43 @@
+    [Fact]
+    public void GrantsAccess()
+    {
+        SecretName secretName = _fixture.SecretToCreateName;
```
Understood, thanks for pointing out the flakiness that was existing in the test suite, have fixed all the occurences by creating test specific resources and have deleted/disposed the secrets/resources created in the same test file.

amanda-tarafa replied:
I think you forgot to change this one.

YashSaraf11 commented on secretmanager/api/SecretManager.Samples.Tests/IamRegionalRevokeAccessTests.cs:
```diff
@@ -0,0 +1,43 @@
+        Policy result = _sample.IamRegionalRevokeAccess(
+          projectId: secretName.ProjectId, locationId: secretName.LocationId, secretId: secretName.SecretId,
+          member: "group:test@google.com");
+        Assert.DoesNotContain(result.Bindings,
```
Have first added the policy to make sure the binding is there and then revoked the access to make it robust.

YashSaraf11 commented on secretmanager/api/SecretManager.Samples.Tests/IamRegionalGrantAccessTests.cs:
```diff
@@ -0,0 +1,43 @@
+    [Fact]
+    public void GrantsAccess()
+    {
+        SecretName secretName = _fixture.SecretToCreateName;
```
Here we are granting the access, so this test verifies that post the call the binding contains the required roles and members.

YashSaraf11 commented on secretmanager/api/SecretManager.Samples/EnableRegionalSecretVersionWithEtag.cs:
```diff
@@ -0,0 +1,51 @@
+        {
+            SecretVersionName = SecretVersionName.FromProjectLocationSecretSecretVersion(
+              projectId, locationId, secretId, secretVersionId
+            ),
```
This is being automatically reformatted whenever `dotnet format` is run, if I manually change then certain times the checks fails whenever the PR is updated.

amanda-tarafa replied:
I'm not sure what's doing that. What IDE are you using and are you sure you don't have default style settings that are more suited for a different language? The .editorconfig at the root of this repo [should be respected](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-format) by `dotnet format` and that definetely wouldn't bring parenthesis to the next line. There are other styling issues around indents etc. that I have flagged in several of the files that do not match what's defined in `.editorconfig` (4 space indents).

And I'm not sure what checks would fail, we certainly don't have any style checks on PR presubmits.

YashSaraf11 commented on secretmanager/api/SecretManager.Samples.Tests/IamRegionalRevokeAccessTests.cs:
```diff
@@ -0,0 +1,43 @@
+    public void GrantsAccess()
+    {
+        SecretName secretName = _fixture.SecretToCreateName;
+        Policy result = _sample.IamRegionalRevokeAccess(
+          projectId: secretName.ProjectId, locationId: secretName.LocationId, secretId: secretName.SecretId,
```
This is auto formatted by `dotnet format`.

amanda-tarafa reviewed (changes requested):
@YashSaraf11 Can you rebase this PR on main? It now has accumulated changes for other products with changes in over 160 files, and it's very hard to review. Thanks!

YashSaraf11 commented:
> @YashSaraf11 Can you rebase this PR on main? It now has accumulated changes for other products with changes in over 160 files, and it's very hard to review. Thanks!

Have rebased this PR on main. Sorry for the inconvenience.

amanda-tarafa commented on secretmanager/api/SecretManager.Samples.Tests/CreateRegionalSecretTests.cs:
```diff
@@ -0,0 +1,46 @@
+    public void CreatesRegionalSecrets()
+    {
+        // Get the SecretName from the set ProjectId & LocationId.
+        SecretName secretName = _fixture.SecretToCreateName;
```
Why do you need a SecretToCreateName here? Can't you just create a secret with a random ID? This makes it look like this secret will be used somewhere else, but that's really not possible because we can't rely on tests executing in any given order.

YashSaraf11 replied:
I will create the secret here separetly as mentioned for tests.

amanda-tarafa replied:
I think you forgot to push a commit. This is not done yet. Also, once you've changed this, it's likely that you can delete SecretToCreateName from the fixture.

amanda-tarafa commented on secretmanager/api/SecretManager.Samples.Tests/EditRegionalSecretAnnotationsTests.cs:
```diff
@@ -0,0 +1,61 @@
+    public void EditRegionalSecretsAnnotations()
+    {
+        // Get the SecretName from the set ProjectId & LocationId.
+        SecretName secretName = _fixture.Secret.SecretName;
```
Just double checking that adding new annotations to this shared secret won't make other tests fail.

YashSaraf11 replied:
No, annotation wouldn't affect other tests that the secret is handling.

amanda-tarafa commented on secretmanager/api/SecretManager.Samples.Tests/IamRegionalGrantAccessTests.cs:
```diff
@@ -0,0 +1,48 @@
+    public void GrantsAccess()
+    {
+        // Get the secret name.
+        SecretName secretName = _fixture.SecretToCreateName;
```
This test will fail if the create secret test happens to run after and not before. We cannot rely on the order in which tests run. Here you need to create a secret specific for this test and delete it at the end, similar to what you have done elsewhere.

YashSaraf11 replied:
Sure, I will create separate secret for these tests as well.

amanda-tarafa commented on secretmanager/api/SecretManager.Samples.Tests/UpdateRegionalSecretWithAliasTests.cs:
```diff
@@ -0,0 +1,48 @@
+    public void UpdatesRegionalSecretsWithAlias()
+    {
+        // Get the secret name.
+        SecretName secretName = _fixture.Secret.SecretName;
```
Double checking here that modifying the shared secret won't make other tests fail.

amanda-tarafa reviewed (changes requested):
This is looking good in terms of possible flakiness because of resource sharing across tests. I think there a couple of tests that still need updating, I've called those out. Thanks!

There are a couple other comments for sample code simplification.

And there are the style comments, about the new-line parenthesis and indents. Those are off. I replied to your comment about `dotnet format`. We should fix these at least in sample code, it happens in some tests as well, because it's really an unfamiliar style.

YashSaraf11 commented:
@amanda-tarafa 
Thanks for the detailed review! I've incorporated your feedback and fixed the dotnet format issues.
I also looked into the secret updates. As you mentioned, they're updating different pieces and don't affect the tests.

I believe I've addressed all your points, but please let me know if anything else comes up.

YashSaraf11 reviewed (commented):
@amanda-tarafa 
Thanks for the detailed review! I've incorporated your feedback and fixed the dotnet format issues.
I also looked into the secret updates. As you mentioned, they're updating different pieces and don't affect the tests.

I believe I've addressed all your points, but please let me know if anything else comes up.

amanda-tarafa reviewed (changes requested):
There's still one of the secret names that is shared when it shouldn't. I've flagged it.

Also, style is still off in some tests. What IDE are you using?

YashSaraf11 commented:
> There's still one of the secret names that is shared when it shouldn't. I've flagged it.
> 
> Also, style is still off in some tests. What IDE are you using?

@amanda-tarafa I will update it, can you mention if the style can be due to different dotnet version being used?

amanda-tarafa reviewed (approved):
Thanks for all the fixes!
